### ISSUE TITLE
enable multi-gpu acceleration for offline validation

### ISSUE THREAD

**Issue by baibaidj** [ultralytics/ultralytics#22567]
### Search before asking

- [x] I have searched the Ultralytics [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

It's great to see that multi-GPU acceleration is now supported for validation during training.
It would be even more beneficial if this capability could be extended to post-training validation as well.

### Use case

This enhancement would allow us to process large-scale inference tasks much faster using multiple GPUs, thereby helping to quickly identify model failure cases and gain insights for further optimization.

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @baibaidj, thank you for your interest in Ultralytics 🚀! This is an automated response — an Ultralytics engineer will follow up soon to assist further.

Your feature request to enable multi-GPU acceleration for offline/post-training validation is noted and appreciated. Extending the current multi-GPU validation capabilities beyond training would indeed help accelerate large-scale evaluation and failure analysis.

To help us scope this enhancement efficiently (and to smooth a potential PR from you 🙌), please share:
- The exact workflow you want to accelerate (e.g., CLI vs Python, single model vs batch of models)
- Expected interface and behavior (e.g., `yolo val ... device=0,1,2,3`, distributed strategy preferences)
- Environment details (OS, CUDA/CUDNN, Python, PyTorch versions, GPU topology)
- Dataset scale and typical batch/image sizes, desired throughput/latency goals
- Any constraints (determinism, reproducibility across ranks, metric aggregation expectations)

We recommend a visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Looking forward to your design details and proposed approach for multi-GPU offline validation. Thanks again for offering to contribute 💪

**Comment by glenn-jocher** [comment]
Thanks, this makes sense and would be useful for large offline evals. The validators already aggregate metrics across ranks, but standalone `val` currently forces the dataloader to `rank=-1`, so only one GPU is used; see [Detection validation reference](https://docs.ultralytics.com/reference/models/yolo/detect/val/#get_dataloader) and the base [Validator flow](https://docs.ultralytics.com/reference/engine/validator/). A minimal PR would change `rank=-1` to `rank=RANK` in each task validator’s `get_dataloader(...)`, e.g.:

```python
# ultralytics/models/yolo/detect/val.py
return build_dataloader(
    dataset, batch_size, self.args.workers,
    shuffle=False, rank=RANK, drop_last=self.args.compile, pin_memory=self.training,
)
```

With that, users can launch multi-GPU offline validation via `torchrun`:

```bash
torchrun --nproc_per_node=2 yolo val model=yolo11n.pt data=coco8.yaml device=0,1 save_json=True
```

Please keep the change scoped (no additions to `default.yaml`), add a brief note to the [Model Validation guide](https://docs.ultralytics.com/modes/val/), and include a simple check that results are consistent across 1 vs N GPUs. Happy to review your PR.

**Comment by baibaidj** [comment]
> Thanks, this makes sense and would be useful for large offline evals. The validators already aggregate metrics across ranks, but standalone `val` currently forces the dataloader to `rank=-1`, so only one GPU is used; see [Detection validation reference](https://docs.ultralytics.com/reference/models/yolo/detect/val/#get_dataloader) and the base [Validator flow](https://docs.ultralytics.com/reference/engine/validator/). A minimal PR would change `rank=-1` to `rank=RANK` in each task validator’s `get_dataloader(...)`, e.g.:
> 
> # ultralytics/models/yolo/detect/val.py
> return build_dataloader(
>     dataset, batch_size, self.args.workers,
>     shuffle=False, rank=RANK, drop_last=self.args.compile, pin_memory=self.training,
> )
> With that, users can launch multi-GPU offline validation via `torchrun`:
> 
> torchrun --nproc_per_node=2 yolo val model=yolo11n.pt data=coco8.yaml device=0,1 save_json=True
> Please keep the change scoped (no additions to `default.yaml`), add a brief note to the [Model Validation guide](https://docs.ultralytics.com/modes/val/), and include a simple check that results are consistent across 1 vs N GPUs. Happy to review your PR.

Sure.  I have submitted a PR #22579  with changes that I thought should be essential, but have not got time to do thorough testing and documentation on usage.  In fact, the PR that I made requires no effort in users part, they just need to specify which gpu devices to do offline validation, the program will automatically run single-gpu or multi-gpu validation based on the specification.  

However, it seems that there is another PR #22569  also addressing this issue and seems to incorporate some of my changes. I guess I should stop here and let  @Y-T-G  finished the rest of this feature?
